README.md
# website-screenshot

[![Rust: 1.61+](https://img.shields.io/badge/rust-1.61+-93450a)](https://blog.rust-lang.org/2022/05/19/Rust-1.61.0.html)
[![Continuous Delivery](https://github.com/devtomio/website-screenshot/actions/workflows/continuous-delivery.yml/badge.svg)](https://github.com/devtomio/website-screenshot/actions/workflows/continuous-delivery.yml)
[![Continuous Integration](https://github.com/devtomio/website-screenshot/actions/workflows/continuous-integration.yml/badge.svg)](https://github.com/devtomio/website-screenshot/actions/workflows/continuous-integration.yml)
[![Integration Tests](https://github.com/devtomio/website-screenshot/actions/workflows/integration-tests.yml/badge.svg)](https://github.com/devtomio/website-screenshot/actions/workflows/integration-tests.yml)

**📸 website screenshots as a service**

## Features

- 💫 powered by [Rust]
- 🚀 blazing fast
- 👮​​​‍‍​ built-in [ratelimiter](https://github.com/antifuchs/governor)
- 👜 built-in [storage providers](#storage-providers)
- 🛡️ built-in [authentication](#authentication)
- 🗼 configurable

## Deployment

### Prerequisites

- [Rust] 1.61 or greater
- [Redis] 6 or greater
- [Chrome] browser
- [Chromedriver] (must match with the version your [Chrome] browser)

### Environment Variables

- `PORT` - the port that the application will run (optional, defaults to `3000`)
- `REDIS_URL` - the address of your redis database (required)
- `FULLSCREEN_SCREENSHOT` - if set, it will screenshot the whole website (optional)
- `CHECK_IF_NSFW` - if set, it will check if the url is marked as NSFW (optional)
- `GOOGLE_CHROME_PATH` - the path to the chrome executable (optional)
- `CHROMEDRIVER_ADDRESS` - the address on where chromedriver is listening (optional, defaults to `http://localhost:9515`)
- `DARK_MODE` - if set, it will take screenshots in dark mode, if the website supports it (optional)
- `FORCE_NSFW_CHECK` - if set, force NSFW check (optional)
- `FORCE_DARK_MODE` - if set, force dark mode (optional)
- `CHROME_FLAGS` - additional flags to provide to chrome (optional, example: `--a,--b,-c`)

### Railway

[![Deploy on Railway](https://railway.app/button.svg)](https://railway.app/new/template/3ZVgSw?referralCode=tomio)

#### Changing Providers

To change the provider in railway:

1. Open the repo that railway made for you.
2. Open the [Dockerfile](./Dockerfile)
3. Edit lines 39 and 44 with `fleet build --release --no-default-features --features <your_provider>`.
4. Deploy your app again.

### Pre-built Binaries

**⚠️- You can't change the provider when using the pre-built binaries.**

#### Linux

```sh
curl -fsSL https://raw.githubusercontent.com/devtomio/website-screenshot/main/install.sh
```

#### Other

See the [Releases page](https://github.com/devtomio/website-screenshot/releases) of this repository and download the appropriate binary.

### Docker

**⚠️- You can't change the provider when using the docker image.**

A dockerized version of this application is available [here](https://github.com/devtomio/website-screenshot/pkgs/container/website-screenshot).

### Building from source

1. Clone this repository. e.g. `git clone https://github.com/devtomio/website-screenshot`
2. Build the binary `cargo build --release`
3. Start chromedriver in the background `chromedriver &`
4. Run the app `./target/release/website-screenshot`

## Authentication

Authentication will be enabled if the `AUTH_TOKEN` variable is set.

It will check if the `Authorization` header sent by the user is equal to the `AUTH_TOKEN` that you set.

## Storage Providers

### Fs (Filesystem) Provider

_📝 - This is the default provider._

#### Build Command

```sh
cargo build --release
```

### [Cloudinary] Provider

_📝 - You must make an unsigned upload preset._

#### Environment Variables

- `CLOUDINARY_API_KEY` - your api key (required)
- `CLOUDINARY_UPLOAD_PRESET` - the name of your unsigned upload preset (required)
- `CLOUDINARY_CLOUD_NAME` - the cloud name that you set during registration (required)

#### Build Command

```sh
cargo build --release --no-default-features --features cloudinary_provider
```

### Amazon AWS [S3] Provider

_⚠️ - This is untested. If you encounter a bug please don't hesitate to open an issue._

#### Environment Variables

- `S3_BUCKET_NAME` - the name of your s3 bucket (required)
- `S3_REGION` - the region of your s3 bucket (required, only optional if using [self-hosted s3](#self-hosted-s3-storage))
- `S3_ACCESS_KEY` - your access key (required)
- `S3_SECRET_KEY` - your secret key (required)
- `S3_SECURITY_TOKEN` - your security token (optional)
- `S3_SESSION_TOKEN` - your session token (optional)

#### Self-hosted S3 Storage

To use a self-hosted S3 Storage, set the `S3_ENDPOINT_URL` environment variable to the endpoint url.

#### Build Command

```sh
cargo build --release --no-default-features --features s3_provider
```

### [Tixte] Provider

#### Environment Variables

- `TIXTE_UPLOAD_KEY` - your upload key, can be found on the integrations tab (required)
- `TIXTE_DOMAIN_CONFIG` - whether to use random domains or a specific domain (required, can only be `standard`, or `random`)
- `TIXTE_CUSTOM_DOMAIN` - the specific domain to use (only required when `TIXTE_DOMAIN_CONFIG` is set to `standard`)

#### Build Command

```sh
cargo build --release --no-default-features --features tixte_provider
```

### [Sled] Provider

#### Environment Variables

- `SLED_PATH` - the path to provide when opening the sled database (optional, defaults to `.website-screenshot`)

#### Build Command

```sh
cargo build --release --no-default-features --features sled_provider
```

## License

Licensed under either of

- Apache License, Version 2.0 ([LICENSE-APACHE](LICENSE-APACHE) or <http://apache.org/licenses/LICENSE-2.0>)
- MIT license ([LICENSE-MIT](LICENSE-MIT) or <http://opensource.org/licenses/MIT>)

### Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall
be dual licensed as above, without any additional terms or conditions.

## Contributors ✨

Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tr>
    <td align="center"><a href="https://tomio.codes/"><img src="https://avatars.githubusercontent.com/u/75403863?v=4?s=100" width="100px;" alt=""/><br /><sub><b>Tomio</b></sub></a><br /><a href="https://github.com/devtomio/website-screenshot/commits?author=devtomio" title="Code">💻</a> <a href="https://github.com/devtomio/website-screenshot/commits?author=devtomio" title="Documentation">📖</a> <a href="#infra-devtomio" title="Infrastructure (Hosting, Build-Tools, etc)">🚇</a> <a href="#maintenance-devtomio" title="Maintenance">🚧</a></td>
  </tr>
</table>

<!-- markdownlint-restore -->
<!-- prettier-ignore-end -->

<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the [all-contributors](https://github.com/all-contributors/all-contributors) specification. Contributions of any kind welcome!

[Rust]: https://rust-lang.org
[Redis]: https://redis.io
[Chrome]: https://google.com/chrome
[Chromedriver]: https://chromedriver.chromium.org
[Cloudinary]: https://cloudinary.com
[S3]: https://aws.amazon.com/s3
[Tixte]: https://tixte.com
[Sled]: https://sled.rs


CHANGELOG.md
# Changelog

## [2.0.0](https://github.com/devtomio/website-screenshot/compare/v1.2.0...v2.0.0) (2023-02-23)


### ⚠ BREAKING CHANGES

* remove cargo releases

### Features

* add ability to add additional chrome flags (closes [#36](https://github.com/devtomio/website-screenshot/issues/36)) ([c1dfb04](https://github.com/devtomio/website-screenshot/commit/c1dfb047fb93628213d3a8b5de8ee904a33c0308))
* add more errors ([bcb7e12](https://github.com/devtomio/website-screenshot/commit/bcb7e12587a395d069bbf99208f717f078aae7a8))
* add width/height to /screenshot POST request ([#38](https://github.com/devtomio/website-screenshot/issues/38)) ([886a21b](https://github.com/devtomio/website-screenshot/commit/886a21b4f01a54531787a8e599ec0ec3bbbf199e))
* change msrv to `1.61` ([#53](https://github.com/devtomio/website-screenshot/issues/53)) ([f226ace](https://github.com/devtomio/website-screenshot/commit/f226acecee02164b9eab40114a5f27db2842a880))


### Bug Fixes

* **deps:** update all non-major dependencies ([#35](https://github.com/devtomio/website-screenshot/issues/35)) ([4888f1d](https://github.com/devtomio/website-screenshot/commit/4888f1da326d9bab230917c4545a997b93b14bc0))
* **deps:** update all non-major dependencies ([#42](https://github.com/devtomio/website-screenshot/issues/42)) ([8778945](https://github.com/devtomio/website-screenshot/commit/87789456f7e8569218e40b04253b2c609139583e))
* **deps:** update all non-major dependencies ([#50](https://github.com/devtomio/website-screenshot/issues/50)) ([552e18e](https://github.com/devtomio/website-screenshot/commit/552e18ec9497d7de0bddef16a3cd32d21392916c))
* **deps:** update rust crate anyhow to 1.0.58 ([#40](https://github.com/devtomio/website-screenshot/issues/40)) ([803069d](https://github.com/devtomio/website-screenshot/commit/803069daac68ac7933069b5215af871e90668b40))
* remove cargo releases ([308332e](https://github.com/devtomio/website-screenshot/commit/308332e9709b75d673b6ebce06bf1d29eaffc785))

## [1.2.0](https://github.com/devtomio/website-screenshot/compare/v1.1.0...v1.2.0) (2022-05-25)


### Features

* allow administrators to force nsfw checks or dark mode ([a6a9ca9](https://github.com/devtomio/website-screenshot/commit/a6a9ca99369adb2131e0002e58da582c10fb678b))
* let the user provide the chromedriver address ([24e82e2](https://github.com/devtomio/website-screenshot/commit/24e82e23253426e8f5a17df8331d3d2183859792))


### Bug Fixes

* **ci:** fix docker image ([c6d92a8](https://github.com/devtomio/website-screenshot/commit/c6d92a8832ded7782769a6ffb2dbbe9b2b76979e))
* **ci:** fix publishing to crates.io ([942c822](https://github.com/devtomio/website-screenshot/commit/942c822d62dbac68924583c8618e2b285ae794bd))

## [1.1.0](https://github.com/devtomio/website-screenshot/compare/v1.0.0...v1.1.0) (2022-05-25)


### Features

* add `Procfile` for heroku ([93d429a](https://github.com/devtomio/website-screenshot/commit/93d429ad1ff8475095f4280a57097aeef00c546e))
* add a self-hosted server on heroku ([ec45896](https://github.com/devtomio/website-screenshot/commit/ec4589644d779d27ffc7bf30cdb74c089164a39d))
* add docs route ([a87ff53](https://github.com/devtomio/website-screenshot/commit/a87ff53d351f343c8646ac223d282f2692f40f3e))
* add dprint checks ([7bb9e7c](https://github.com/devtomio/website-screenshot/commit/7bb9e7c2ab4e78c28f0dba1a7bca352662644d56))
* add more checks to prevent mismatch ([1a11545](https://github.com/devtomio/website-screenshot/commit/1a11545c02ac1d8d820321c0d5550d05f159694d))
* add nsfw checks ([bcc8360](https://github.com/devtomio/website-screenshot/commit/bcc83606756e226937beacc956ce4a309e1db5ab))
* **ci:** cache deps ([931844a](https://github.com/devtomio/website-screenshot/commit/931844a9bc7c05785bc99572147bce1118fffa3d))
* dark mode ([0ca4555](https://github.com/devtomio/website-screenshot/commit/0ca4555a5bc94c79e5a2af940f1dff7daf671a54))
* prevent bot detection ([39f982d](https://github.com/devtomio/website-screenshot/commit/39f982dd6fc3924df36276c5bc180740f7013530))
* remove use of `portpicker` ([b58415e](https://github.com/devtomio/website-screenshot/commit/b58415e8b6206172b2b0ab5afa99ee09050ab2ad))
* **tests:** add basic tests ([42ce400](https://github.com/devtomio/website-screenshot/commit/42ce40002665664a76676966c8d08921a9b3b1f8))
* use docker for heroku ([6a33b8b](https://github.com/devtomio/website-screenshot/commit/6a33b8bb70bf128af074bda676f1791537869dba))


### Bug Fixes

* add whitelisted ips option ([cc6d626](https://github.com/devtomio/website-screenshot/commit/cc6d626a8e1f0f7eb87639c0cb55feef19036763))
* **deps:** update rust crate once_cell to 1.12.0 ([#27](https://github.com/devtomio/website-screenshot/issues/27)) ([96e6333](https://github.com/devtomio/website-screenshot/commit/96e6333ed89de3a89e317899cf5ea3490ff5f2d5))
* **deps:** update rust crate rayon to 1.5.3 ([#24](https://github.com/devtomio/website-screenshot/issues/24)) ([787d4e4](https://github.com/devtomio/website-screenshot/commit/787d4e4c308b6c81bc51708616e9741ff6eef858))
* **deps:** update rust crate tokio to 1.18.2 ([#20](https://github.com/devtomio/website-screenshot/issues/20)) ([76d20ac](https://github.com/devtomio/website-screenshot/commit/76d20ac893eebea36628cd140225327e9cb49fbd))
* derive `Eq` for `Error` ([4efd1ab](https://github.com/devtomio/website-screenshot/commit/4efd1ab757bcdc4806b431ee9588a648159538ac))
* docker build ([0abd36e](https://github.com/devtomio/website-screenshot/commit/0abd36e92a6ff417ae711ad888599c1c8ffafc50))
* heroku build ([14a7b12](https://github.com/devtomio/website-screenshot/commit/14a7b1286b4f466d0a2503bac8a76c0dd6626bf0))
* hide scrollbar ([72c38e5](https://github.com/devtomio/website-screenshot/commit/72c38e5d099d2a470081d2a63a3d19778afb6774))
* remove heroku ([a5b20d8](https://github.com/devtomio/website-screenshot/commit/a5b20d8843034460711c8510bf2ad3831b9999a5))
* set remote debugging address ([410bb07](https://github.com/devtomio/website-screenshot/commit/410bb078ff7150d2ba704f78add1b9896b4639a2))

src/main.rs
#![warn(clippy::pedantic)]
#![allow(
    clippy::unreadable_literal,
    clippy::module_name_repetitions,
    clippy::unused_async,
    clippy::too_many_lines,
    clippy::cast_possible_truncation,
    clippy::missing_errors_doc,
    clippy::must_use_candidate,
    clippy::wildcard_imports
)]

#[macro_use]
extern crate tracing;

use std::borrow::ToOwned;
use std::env;
use std::sync::Arc;

use actix_web::middleware::Compress;
use actix_web::{web, App, Error, HttpServer};
use actix_web_static_files::ResourceFiles;
use cdp::ChromeCommand;
use evasions::{
    evaluate_on_new_document,
    CHROME_APP,
    CHROME_CSI,
    CHROME_LOADTIMES,
    CHROME_RUNTIME,
    IFRAME_CONTENTWINDOW,
    MEDIA_CODECS,
    NAVIGATOR_HARDWARECONCURRENCY,
    NAVIGATOR_LANGUAGES,
    NAVIGATOR_PERMISSIONS,
    NAVIGATOR_PLUGINS,
    NAVIGATOR_VENDOR,
    NAVIGATOR_WEBDRIVER,
    UTILS,
    WEBGL_VENDOR,
    WINDOW_OUTERDIMENSIONS,
};
use fantoccini::{Client, ClientBuilder};
use providers::{Provider, Storage};
use reqwest::Client as ReqwestClient;
use serde_json::{Map, Value};
use tracing_actix_web::TracingLogger;
use util::{initialize_tracing, load_env};
use website_screenshot_actix_governor::{Governor, GovernorConfigBuilder};

pub mod cdp;
pub mod error;
pub mod evasions;
pub mod middlewares;
pub mod pattern;
pub mod providers;
pub mod routes;
pub mod util;

pub type Result<T, E = Error> = anyhow::Result<T, E>;

#[derive(Debug)]
pub struct State {
    pub browser: Arc<Client>,
    pub storage: Arc<Storage>,
    pub reqwest: ReqwestClient,
}

include!(concat!(env!("OUT_DIR"), "/generated.rs"));

#[actix_web::main]
async fn main() -> anyhow::Result<()> {
    load_env();
    initialize_tracing();

    let mut capabilities = Map::new();
    let mut args = vec![
        "--disable-gpu".to_owned(),
        "--no-sandbox".to_owned(),
        "--disable-dev-shm-usage".to_owned(),
        "--headless".to_owned(),
        "--hide-scrollbars".to_owned(),
        "--whitelisted-ips=".to_owned(),
    ];

    if let Ok(flags) = env::var("CHROME_FLAGS") {
        let flags = flags.split(',').map(ToOwned::to_owned).collect::<Vec<_>>();

        args.extend_from_slice(&flags);
    };

    let chrome_opts = match env::var("GOOGLE_CHROME_PATH") {
        Ok(path) => serde_json::json!({
            "binary": path,
            "args": args
        }),
        Err(_) => serde_json::json!({ "args": args }),
    };

    capabilities.insert("goog:chromeOptions".to_owned(), chrome_opts);

    let chromedriver_address = match env::var("CHROMEDRIVER_ADDRESS") {
        Ok(address) => address,
        Err(_) => "http://localhost:9515".to_owned(),
    };

    let client =
        ClientBuilder::rustls().capabilities(capabilities).connect(&chromedriver_address).await?;

    info!("Connected to chromedriver at localhost:9515");

    // To hide headless nature (for Cloudflare, etc.)
    tokio::join!(
        evaluate_on_new_document(&client, UTILS, vec![]),
        evaluate_on_new_document(&client, CHROME_APP, vec![]),
        evaluate_on_new_document(&client, CHROME_CSI, vec![]),
        evaluate_on_new_document(&client, CHROME_LOADTIMES, vec![]),
        evaluate_on_new_document(&client, CHROME_RUNTIME, vec![Value::Bool(false)]),
        evaluate_on_new_document(&client, IFRAME_CONTENTWINDOW, vec![]),
        evaluate_on_new_document(&client, MEDIA_CODECS, vec![]),
        evaluate_on_new_document(&client, NAVIGATOR_HARDWARECONCURRENCY, vec![Value::Number(
            4.into()
        )]),
        evaluate_on_new_document(&client, NAVIGATOR_LANGUAGES, vec![Value::Array(vec![
            "en-US".into(),
            "en".into()
        ])]),
        evaluate_on_new_document(&client, NAVIGATOR_PERMISSIONS, vec![]),
        evaluate_on_new_document(&client, NAVIGATOR_PLUGINS, vec![]),
        evaluate_on_new_document(&client, NAVIGATOR_VENDOR, vec![Value::String(
            "Google Inc.".to_owned()
        )]),
        evaluate_on_new_document(&client, NAVIGATOR_WEBDRIVER, vec![]),
        evaluate_on_new_document(&client, WEBGL_VENDOR, vec![
            Value::String("Intel Inc.".to_owned()),
            Value::String("Intel Iris OpenGL Engine".to_owned())
        ]),
        evaluate_on_new_document(&client, WINDOW_OUTERDIMENSIONS, vec![]),
    );

    // Override user-agent
    let user_agent = client
        .issue_cmd(ChromeCommand::ExecuteCdpCommand(
            "Browser.getVersion".to_owned(),
            serde_json::json!({}),
        ))
        .await
        .expect("Failed issuing cmd")["userAgent"]
        .as_str()
        .expect("Failed to get user agent")
        .to_owned();

    let new_user_agent = user_agent.replace("HeadlessChrome", "Chrome");

    client
        .issue_cmd(ChromeCommand::ExecuteCdpCommand(
            "Network.setUserAgentOverride".to_owned(),
            serde_json::json!({
                "userAgent": new_user_agent,
                "acceptLanguage": "en-US,en"
            }),
        ))
        .await
        .expect("Failed issuing cmd");

    let governor_config = GovernorConfigBuilder::default()
        .per_second(60)
        .burst_size(20)
        .with_headers()
        .finish()
        .expect("Failed to build ratelimiter");

    let state = web::Data::new(State {
        browser: Arc::new(client.clone()),
        storage: Arc::new(Storage::new()),
        reqwest: ReqwestClient::new(),
    });

    let port =
        env::var("PORT").map(|p| p.parse::<u16>().expect("Failed to parse port")).unwrap_or(3000);

    info!("Server listening at localhost:{}", port);

    HttpServer::new(move || {
        let static_files = generate();

        App::new()
            .wrap(Compress::default())
            .wrap(middlewares::Auth)
            .wrap(TracingLogger::default())
            .wrap(Governor::new(&governor_config))
            .app_data(state.clone())
            .service(routes::screenshot_route)
            .service(routes::get_screenshot)
            .service(routes::schema_route)
            .service(routes::index_route)
            .service(ResourceFiles::new("/", static_files))
    })
    .bind(("0.0.0.0", port))?
    .run()
    .await?;

    client.close().await?;

    Ok(())
}
